src/agent.py

```diff
@@ -104,7 +104,8 @@ def agent_discovery_url(self) -> str:
             unit_ip = str(binding.network.bind_address)
             try:
                 ipaddress.ip_address(unit_ip)
-                return f"http://{unit_ip}:{jenkins.WEB_PORT}"
+                env_dict = typing.cast(typing.Dict[str, str], self.jenkins.environment)
```

Comment:
Very nitpicky (no need to implement here), perhaps the typing.cast can go under jenkins.Jenkins once so that others needing the environment can automatically infer it. Also perhaps a typed dict can help future developers know what the env contains :)

Reply:
good point, I'll address this in a follow-up PR, alongside with the API token generation and tests refactoring